src/polyfillstub/intersection-observer-stub.js

```diff
@@ -27,6 +27,12 @@ import {Services} from '../services';
 
 const UPGRADERS = '_upgraders';
 
+/** @type {IntersectionObserver} */
+let NativeIntersectionObserver;
```

Comment:
This is dangerous, because both v0.js and amp-intersection-observer.js independently compile/bundle this module. So these vars will be different in both cases. If we won't run into a bug now, we likely will do later. In this case, it might be better to hoist private props on the window object.

Overall I'm wondering if a modified previous concept that you had could be the clearest here. Maybe something like this:

```
// polyfills/in-ob.js:

export function install(win) {
  const Native = win.IntersectionObserver;
  if (!Native || !supportsRootMargin(Native)) {
    win.IntersectionObserver = createStub(Native);
  }
  fixEntry(win);
}
```

```
// polyfillstubs/in-ob.js

/** @return {typeof IntersectionObserver} */
export function createStub(nativeClass) {
  if (nativeClass) {
    const dispatcher = function(callback, opts) {
      if (opts && opts.root && opts.root.nodeType == 9) {
        return new IntersectionObserverStub(callback, opts);
      }
      return new nativeClass(callback, opts);
    };
    dispatcher[STUB] = IntersectionObserverStub;
    dispatcher[NATIVE] = nativeClass;
    return dispatcher;
  }
  return IntersectionObserverStub;
}
```

And then the upgrade to polyfill will look essentially the same as you had before.

WDYT?

Reply:
Is the benefit to this approach vs. hoisting to `window` that we don't need to pollute window at all?

Reply:
Yes. If we can avoid easily we should. But here, we also have several structures that need to be preserved. It'd be especially not great to hoist many state props on the window object.

Reply:
This is not a hard-and-fast rule, of course. But if there's a clear and relevant non-window place where to put this state, I think we should do so.

Reply:
I _think_ I've incorporated your feedback into the latest push. LMK if I'm missing anything!